osu.Game/Rulesets/Filter/FilterCriteria_QueryParser.cs

```diff
@@ -6,75 +6,121 @@
 using System.Text.RegularExpressions;
 using osu.Game.Beatmaps;
 
-namespace osu.Game.Screens.Select
+namespace osu.Game.Rulesets.Filter
 {
-    internal static class FilterQueryParser
+    public partial class FilterCriteria
```

Comment:
Making this `partial` is maybe a bit controversial. I feel that:

* a second class for this is overkill
* it'll be nicer and easier on the implementors to have it in one place
* the parsing part is mostly separate logic from the interpretation
* I wanted to avoid 500-line files

I'll split this back to a separate class, or merge the files, on request.